src/libxtp/sternheimerw.cc

```diff
@@ -0,0 +1,394 @@
+
+  this->_num_occ_lvls = _orbitals.getNumberOfAlphaElectrons();
+  this->_basis_size = _orbitals.getBasisSetSize();
+  this->_Hamiltonian_Matrix = Hamiltonian();
+  this->_density_Matrix = _orbitals.DensityMatrixGroundState();
+  this->_overlap_Matrix = OverlapMatrix();
+  this->_mo_coefficients = _orbitals.MOCoefficients();
+  this->_mo_energies = _orbitals.MOEnergies();
+  this->_inverse_overlap = _overlap_Matrix.inverse();
+}
+
+void SternheimerW::initializeMultishift(int size) { 
+  _multishift.setMatrixSize(size);
+}
+void SternheimerW::initializePade(int size) {
+  _pade.initialize(size);
+}
+
+Eigen::MatrixXd SternheimerW::OverlapMatrix() const{
+
+  AOBasis basis = _orbitals.SetupDftBasis();
+  AOOverlap overlap;
+  overlap.Fill(basis);
+  return overlap.Matrix();
+}
+
+Eigen::MatrixXd SternheimerW::DensityMatrix() const{
+
+  return _orbitals.DensityMatrixGroundState();
+}
+
+Eigen::MatrixXd SternheimerW::Hamiltonian() const{
+
+  const Eigen::MatrixXd& mo_coefficients = _orbitals.MOCoefficients();
+  const Eigen::MatrixXd& mo_energies = _orbitals.MOEnergies().asDiagonal();
+  const Eigen::MatrixXd overlap = OverlapMatrix();
+  Eigen::MatrixXd H = overlap * mo_coefficients * mo_energies *
+                      mo_coefficients.transpose() * overlap;
+
+  return H;
+}
+Eigen::MatrixXcd SternheimerW::CoulombMatrix(Eigen::Vector3d gridpoint) const{
+
+  AOBasis basis = _orbitals.SetupDftBasis();
+  AOESP aoesp;
+  aoesp.setPosition(gridpoint);
+  aoesp.Fill(basis);
+  return aoesp.Matrix();
+}
+
+Eigen::MatrixXcd SternheimerW::SternheimerLHS(const Eigen::MatrixXcd& hamiltonian,
+                                              const Eigen::MatrixXcd& inverse_overlap,
+                                              double eps,
+                                              std::complex<double> omega,
+                                              bool pm) const{
+
+  // distinguish between +w and -w
+  std::complex<double> temp=eps+omega;
+  if (pm != true) {
+    temp = (eps - omega);
+  }
+  return (inverse_overlap*hamiltonian - temp*Eigen::MatrixXcd::Identity(_basis_size,_basis_size));
+}
+
+Eigen::VectorXcd SternheimerW::SternheimerRHS(const Eigen::MatrixXcd& inverse_overlap,
+                                              const Eigen::MatrixXcd& density,
+                                              const Eigen::MatrixXcd& pertubation,
+                                              const Eigen::VectorXcd& coeff) const{
+
+  return -1*(inverse_overlap-density)*pertubation*coeff;
+}
+
+Eigen::VectorXcd SternheimerW::SternheimerSolve(const Eigen::MatrixXcd& LHS,
+                                                const Eigen::VectorXcd& RHS){
+  return LHS.colPivHouseholderQr().solve(RHS);
+}
+
+std::vector<Eigen::MatrixXcd> SternheimerW::DeltaNOneShot(
+    std::vector<std::complex<double>> w, Eigen::Vector3d r) const{
+
+  Eigen::MatrixXcd V = CoulombMatrix(r);
+  
+  double alpha = 2 * (_mo_energies(_num_occ_lvls) - _mo_energies(2));
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Variable 'alpha' is assigned a value that is never used.](https://app.codacy.com/app/votca-package/xtp/pullRequest?prid=4286864)